docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,86 @@
+# Provider agnostic DNS Health checks
+
+## Introduction
+
+The MGC has the ability of extending the DNS configuration of the gateway with the DNSPolicy resource. This resource allows users to configure health checks. As a result of configuring health checks, the controller creates the health checks in Route53, attaching them to the related DNS records. This has the benefit of automatically disabling an endpoint if it becomes unhealthy, and enabling it again when it becomes healthy again.
+This feature has a few shortfalls:
+
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* Ability for the health checks to reach endpoints in private networks
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+* Transparently keep support for other health check providers like Route53
+
+## Proposal
+
+### Control plane
+
+Add a field in the health check section of the DNSPolicy to select the health
+check type to use
+
+```yaml
+healthCheck:
+  # ...
+  port: "..."
+  # ...
+  type: "Provider | Agent"
```

Comment:
Just a thought here. 
The provider vs. the agent are implementations of health checks.
From a high level functionality point of view they are doing the same thing.
However, there is a subtle difference that makes me wonder if the type of health check should be based on the location of the health check.
For example, `internal` vs `external`. Or `private` vs `public`.
If you're using mgc with route53 you can define both types of health checks.
If not, only internal/private is allowed

Reply:
how much does the end user care about this detail I wonder? I want a health check and I want it to remove unhealthy endpoints / alert me when things are going wrong. Not sure how much I care where that health check is?

Reply:
I am wondering basically if we have an agent based approach, is there any advantage to using the cloud providers health checks?
They cost more and work in less scenerios